event-sourcing/event-source-api/src/main/java/uk/gov/justice/services/eventsourcing/source/api/service/EventsPageService.java

```diff
@@ -0,0 +1,106 @@
+                                       final Direction direction,
+                                       final long pageSize,
+                                       final UriInfo uriInfo,
+                                       final UUID streamId) throws SQLException, MalformedURLException {
+
+        final Position positionOf = position(position);
+
+        final List<EventEntry> entities = service.events(positionOf, direction, pageSize, streamId);
+
+        return feedWithLinks(positionOf, pageSize, uriInfo, streamId, entities);
+    }
+
+    private Page<EventEntry> feedWithLinks(final Position position, final long pageSize, final UriInfo uriInfo,
+                                           final UUID streamId, final List<EventEntry> feedEntries) throws MalformedURLException {
+
+        if (feedEntries.isEmpty()) {
+            return new Page<>(new ArrayList<>(),
+                    new PagingLinks(null, null, pageHref(head(), BACKWARD, pageSize, uriInfo),
+                            pageHref(first(), FORWARD, pageSize, uriInfo)));
+        }
+
+        final long maxSequenceId = max(feedEntries);
+        final long minSequenceId = min(feedEntries);
+
+        final boolean newEventsAvailable = !position.isHead() && service.recordExists(maxSequenceId + 1, streamId);
+        final boolean olderEventsAvailable = !position.isFirst() && service.recordExists(minSequenceId - 1, streamId);
+
+        return new Page(feedEntries, new PagingLinks(
+                newEventsAvailable ? pageHref(position(maxSequenceId + 1), FORWARD, pageSize, uriInfo) : null,
+                olderEventsAvailable ? pageHref(position(minSequenceId - 1), BACKWARD, pageSize, uriInfo) : null,
+                pageHref(head(), BACKWARD, pageSize, uriInfo),
+                pageHref(first(), FORWARD, pageSize, uriInfo)));
+    }
+
+    private URL pageHref(final Position position,
+                         final Direction direction,
+                         final long pageSize,
+                         final UriInfo uriInfo) throws MalformedURLException {
+
+        return new URL(uriInfo.getBaseUriBuilder()
+                .path(uriInfo.getPathSegments().get(PATH_SEGMENT_EVENT_STREAMS).getPath())
+                .path(uriInfo.getPathSegments().get(PATH_SEGMENT_EVENT_STREAM_ID).getPath())
+                .path(position.getPosition())
+                .path(deriveDirection(position, direction))
+                .path(String.valueOf(pageSize)).build()
+                .toString());
+    }
+
+    private String deriveDirection(final Position position,
+                                   final Direction direction) {
+
+        final Direction firstDirection = position.isFirst() ? FORWARD : direction;
+        final Direction edgeDirection = position.getPosition().equals(FixedPosition.FIRST.getPosition()) ? FORWARD : firstDirection;
+        return position.isHead() ? BACKWARD.toString() : edgeDirection.toString();
```

Comment:
isn't the first part of the second line (edgeDirection) the same as the first line? 

Also this seems a little confusing, is it possible to make it cleaner? either use named method or temporary booleans to clarify.

Reply:
is deriveDirection clear enough, could it be deriveLink or URL direction?

Reply:
fixed